File: C++ Examples/RedefineAuth.cpp
#include "RedefineAuth.h"
#include <curl/curl.h>
#include <sstream>
#include <cstring>
#include <iostream>
#include <iomanip>
#include <openssl/sha.h>

// Callback function to write response data
static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* response) {
    size_t totalSize = size * nmemb;
    response->append((char*)contents, totalSize);
    return totalSize;
}

// Implementation class
class RedefineAuthClientImpl {
public:
    std::string redefineId;
    std::string baseUrl;
    std::string appVersion;
    
    RedefineAuthClientImpl(const std::string& id, const std::string& url, const std::string& version)
        : redefineId(id), baseUrl(url), appVersion(version) {
        // Initialize curl globally (should be done once in the application)
        curl_global_init(CURL_GLOBAL_DEFAULT);
    }
    
    ~RedefineAuthClientImpl() {
        // Cleanup curl globally (should be done once in the application)
        curl_global_cleanup();
    }
    
    // Generate HWID based on machine info
    std::string getHardwareId() {
        // Get machine info (simplified for example)
        std::string machineInfo = "machine_info_placeholder"; // In real implementation, get actual machine info
        
        // Create SHA256 hash
        unsigned char hash[SHA256_DIGEST_LENGTH];
        SHA256_CTX sha256;
        SHA256_Init(&sha256);
        SHA256_Update(&sha256, machineInfo.c_str(), machineInfo.size());
        SHA256_Final(hash, &sha256);
        
        // Convert to hex string
        std::stringstream ss;
        for(int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
            ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
        }
        
        return ss.str();
    }
    
    // Make HTTP POST request
    AuthResult makeRequest(const std::string& endpoint, const std::string& jsonData) {
        CURL* curl;
        CURLcode res;
        std::string response;
        
        curl = curl_easy_init();
        if(curl) {
            std::string url = baseUrl + endpoint;
            
            // Set URL
            curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
            
            // Set headers
            struct curl_slist* headers = NULL;
            headers = curl_slist_append(headers, "Content-Type: application/json");
            curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
            
            // Set POST data
            curl_easy_setopt(curl, CURLOPT_POSTFIELDS, jsonData.c_str());
            
            // Set callback function to capture response
            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
            
            // Perform the request
            res = curl_easy_perform(curl);
            
            // Get HTTP response code
            long responseCode;
            curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &responseCode);
            
            // Cleanup
            curl_slist_free_all(headers);
            curl_easy_cleanup(curl);
            
            // Handle response
            if(res == CURLE_OK) {
                if(responseCode == 200) {
                    return AuthResult(true, "Operation successful");
                } else {
                    return handleErrorResponse(responseCode, response);
                }
            } else {
                return AuthResult(false, "Network error: " + std::string(curl_easy_strerror(res)));
            }
        }
        
        return AuthResult(false, "Failed to initialize CURL");
    }
    
    // Handle different HTTP error responses
    AuthResult handleErrorResponse(long responseCode, const std::string& responseContent) {
        switch(responseCode) {
            case 503:
                return AuthResult(false, "Application Paused: " + responseContent);
            case 400:
                return AuthResult(false, "Invalid Request: " + responseContent);
            case 404:
                return AuthResult(false, "Resource Not Found: " + responseContent);
            case 403:
                return AuthResult(false, "Access Denied: " + responseContent);
            default:
                return AuthResult(false, "HTTP " + std::to_string(responseCode) + ": " + responseContent);
        }
    }
};

// RedefineAuthClient implementation
RedefineAuthClient::RedefineAuthClient(const std::string& redefineId, 
                                     const std::string& baseUrl,
                                     const std::string& appVersion)
    : impl(new RedefineAuthClientImpl(redefineId, baseUrl, appVersion)) {}

RedefineAuthClient::~RedefineAuthClient() = default;

AuthResult RedefineAuthClient::registerUser(const std::string& username, const std::string& licenseKey) {
    try {
        // Generate HWID
        std::string hwid = impl->getHardwareId();
        
        // Prepare JSON data
        std::string jsonData = "{"
            "\"username\":\"" + username + "\","
            "\"licenseKey\":\"" + licenseKey + "\","
            "\"hwid\":\"" + hwid + "\","
            "\"redefineDeveloperId\":\"" + impl->redefineId + "\","
            "\"appVersion\":\"" + impl->appVersion + "\""
            "}";
        
        // Make request
        return impl->makeRequest("/api/client-users", jsonData);
    } catch(const std::exception& e) {
        return AuthResult(false, "Registration error: " + std::string(e.what()));
    }
}

AuthResult RedefineAuthClient::loginUser(const std::string& username) {
    try {
        // Generate HWID
        std::string hwid = impl->getHardwareId();
        
        // Prepare JSON data
        std::string jsonData = "{"
            "\"username\":\"" + username + "\","
            "\"hwid\":\"" + hwid + "\","
            "\"redefineDeveloperId\":\"" + impl->redefineId + "\","
            "\"appVersion\":\"" + impl->appVersion + "\""
            "}";
        
        // Make request
        return impl->makeRequest("/api/client-users/login", jsonData);
    } catch(const std::exception& e) {
        return AuthResult(false, "Login error: " + std::string(e.what()));
    }
}

File: Java Examples/RedefineAuth.java
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.concurrent.CompletableFuture;
import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import com.google.gson.Gson;
import com.google.gson.JsonObject;

public class RedefineAuth {
    private final String redefineId;
    private final String baseUrl;
    private final String appVersion;
    private final HttpClient httpClient;
    private final Gson gson;

    /**
     * Initialize the RedefineAuth client.
     * 
     * @param redefineId Your Redefine ID
     * @param baseUrl Base URL of your RedefineAuth deployment (default: "https://redefine-auth-v2.vercel.app/")
     * @param appVersion Version of your application (default: "1.0.0.0")
     */
    public RedefineAuth(String redefineId, String baseUrl, String appVersion) {
        this.redefineId = redefineId;
        this.baseUrl = baseUrl.replaceAll("/$", ""); // Remove trailing slash if present
        this.appVersion = appVersion;
        this.httpClient = HttpClient.newHttpClient();
        this.gson = new Gson();
    }

    /**
     * Constructor with default base URL and app version.
     * 
     * @param redefineId Your Redefine ID
     */
    public RedefineAuth(String redefineId) {
        this(redefineId, "https://redefine-auth-v2.vercel.app/", "1.0.0.0");
    }

    /**
     * Register a new user with username and license key.
     * 
     * @param username The username to register
     * @param licenseKey The license key provided by the admin
     * @return AuthResult with success status and message
     */
    public AuthResult register(String username, String licenseKey) {
        try {
            // Generate HWID based on machine info
            String hwid = getHardwareId();
            
            // Prepare request data
            JsonObject requestData = new JsonObject();
            requestData.addProperty("username", username);
            requestData.addProperty("licenseKey", licenseKey);
            requestData.addProperty("hwid", hwid);
            requestData.addProperty("redefineDeveloperId", redefineId);
            requestData.addProperty("appVersion", appVersion);
            
            // Make HTTP POST request
            HttpResponse<String> response = makeRequest("/api/client-users", requestData.toString());
            
            // Handle response
            if (response.statusCode() == 200) {
                return new AuthResult(true, "User registered successfully");
            } else {
                return handleErrorResponse(response);
            }
        } catch (Exception e) {
            return new AuthResult(false, "Registration error: " + e.getMessage());
        }
    }

    /**
     * Login an existing user with username.
     * 
     * @param username The username to login
     * @return AuthResult with success status and message
     */
    public AuthResult login(String username) {
        try {
            // Generate HWID based on machine info
            String hwid = getHardwareId();
            
            // Prepare request data
            JsonObject requestData = new JsonObject();
            requestData.addProperty("username", username);
            requestData.addProperty("hwid", hwid);
            requestData.addProperty("redefineDeveloperId", redefineId);
            requestData.addProperty("appVersion", appVersion);
            
            // Make HTTP POST request
            HttpResponse<String> response = makeRequest("/api/client-users/login", requestData.toString());
            
            // Handle response
            if (response.statusCode() == 200) {
                return new AuthResult(true, "Login successful");
            } else {
                return handleErrorResponse(response);
            }
        } catch (Exception e) {
            return new AuthResult(false, "Login error: " + e.getMessage());
        }
    }

    /**
     * Generate a simple HWID based on machine name and platform info.
     * 
     * @return Hex encoded HWID string
     */
    private String getHardwareId() throws NoSuchAlgorithmException, UnknownHostException {
        // Get machine info
        String machineInfo = InetAddress.getLocalHost().getHostName() + 
                            System.getProperty("os.name") + 
                            System.getProperty("os.arch");
        
        // Create SHA256 hash and encode as hex
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hash = digest.digest(machineInfo.getBytes(StandardCharsets.UTF_8));
        
        // Convert byte array to hex string
        StringBuilder hexString = new StringBuilder();
        for (byte b : hash) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }
        
        return hexString.toString();
    }

    /**
     * Make HTTP POST request to the API.
     * 
     * @param endpoint API endpoint
     * @param jsonData Request data as JSON string
     * @return HttpResponse with status code and body
     * @throws IOException
     * @throws InterruptedException
     */
    private HttpResponse<String> makeRequest(String endpoint, String jsonData) 
            throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + endpoint))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonData))
                .build();
                
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    }

    /**
     * Handle different HTTP error responses.
     * 
     * @param response HTTP response
     * @return AuthResult with appropriate error message
     */
    private AuthResult handleErrorResponse(HttpResponse<String> response) {
        String responseContent = response.body() != null ? response.body() : "Unknown error";
        
        // Handle different status codes
        switch (response.statusCode()) {
            case 503:
                return new AuthResult(false, "Application Paused: " + responseContent);
            case 400:
                return new AuthResult(false, "Invalid Request: " + responseContent);
            case 404:
                return new AuthResult(false, "Resource Not Found: " + responseContent);
            case 403:
                return new AuthResult(false, "Access Denied: " + responseContent);
            default:
                return new AuthResult(false, "HTTP " + response.statusCode() + ": " + responseContent);
        }
    }

    /**
     * AuthResult class to represent the result of authentication operations.
     */
    public static class AuthResult {
        public final boolean success;
        public final String message;
        
        public AuthResult(boolean success, String message) {
            this.success = success;
            this.message = message;
        }
    }

    /**
     * Example usage
     */
    public static void main(String[] args) {
        // Initialize the client
        RedefineAuth client = new RedefineAuth(
            "YOUR_REDEFINE_ID_HERE",           // Change this to your Redefine ID
            "https://redefine-auth-v2.vercel.app/",          // Change this to your deployed RedefineAuth URL
            "1.0.0.0"                         // Change this to match your application version
        );
        
        // Example registration
        System.out.println("Registering user...");
        AuthResult result = client.register("testuser", "LICENSE-KEY-123");
        System.out.println("Registration result: " + (result.success ? "SUCCESS" : "FAILED") + 
                          " - " + result.message);
        
        // Example login
        System.out.println("\nLogging in user...");
        result = client.login("testuser");
        System.out.println("Login result: " + (result.success ? "SUCCESS" : "FAILED") + 
                          " - " + result.message);
    }
}
